File: app/types/query.ts
import type { Column } from './database-structure'

/**
 * クエリモデル型定義
 */

/**
 * 選択されたテーブル（テーブルカード用）
 */
export interface SelectedTable {
  /** 一意識別子（schema.table_name） */
  id: string
  /** スキーマ名 */
  schema: string
  /** テーブル名 */
  name: string
  /** エイリアス */
  alias: string
  /** カラム一覧 */
  columns: Column[]
}

/**
 * WHERE演算子
 */
export type WhereOperator =
  | '='
  | '!='
  | '>'
  | '>='
  | '<'
  | '<='
  | 'LIKE'
  | 'NOT LIKE'
  | 'IN'
  | 'NOT IN'
  | 'BETWEEN'
  | 'IS NULL'
  | 'IS NOT NULL'

/**
 * WHERE条件
 */
export interface WhereCondition {
  id: string
  type: 'condition'
  column: {
    tableAlias: string
    columnName: string
  } | null
  operator: WhereOperator
  value: string | string[] | { from: string; to: string }
  isValid: boolean
}

/**
 * 条件グループ
 */
export interface ConditionGroup {
  id: string
  type: 'group'
  logic: 'AND' | 'OR'
  conditions: Array<WhereCondition | ConditionGroup>
}

/**
 * 選択されたカラム（カラム選択UI用）
 */
export interface SelectedColumn {
  /** テーブルID（SelectedTable.idに対応） */
  tableId: string
  /** テーブルエイリアス */
  tableAlias: string
  /** カラム名 */
  columnName: string
  /** カラムエイリアス */
  columnAlias: string | null
  /** データ型 */
  dataType: string
}

/**
 * テーブル情報
 */
export interface TableInfo {
  /** スキーマ名 */
  schema: string;
  /** テーブル名 */
  name: string;
  /** エイリアス */
  alias?: string;
}

/**
 * JOIN情報
 */
export interface JoinInfo {
  /** JOIN種別 */
  type: 'INNER' | 'LEFT' | 'RIGHT' | 'FULL' | 'CROSS';
  /** 結合先テーブル */
  table: TableInfo;
  /** ON条件 */
  on?: {
    left: string;
    right: string;
    operator: string;
  };
}

/**
 * カラム情報
 */
export interface ColumnInfo {
  /** テーブル */
  table: string;
  /** カラム名 */
  name: string;
  /** エイリアス */
  alias?: string;
  /** 集約関数 */
  aggregate?: 'COUNT' | 'SUM' | 'AVG' | 'MIN' | 'MAX';
}



/**
 * ORDER BY情報
 */
export interface OrderByInfo {
  /** カラム */
  column: string;
  /** 順序 */
  direction: 'ASC' | 'DESC';
}

/**
 * クエリモデル
 */
export interface QueryModel {
  /** テーブル一覧 */
  tables: TableInfo[];
  /** JOIN一覧 */
  joins: JoinInfo[];
  /** カラム一覧 */
  columns: ColumnInfo[];
  /** WHERE条件一覧 (ルートはグループまたは条件のリスト) */
  where: Array<WhereCondition | ConditionGroup>;
  /** GROUP BY一覧 */
  groupBy: string[];
  /** HAVING条件一覧 */
  having: Array<WhereCondition | ConditionGroup>;
  /** ORDER BY一覧 */
  orderBy: OrderByInfo[];
  /** LIMIT */
  limit: number | null;
  /** OFFSET */
  offset: number | null;
}

/**
 * クエリ情報
 */
export interface QueryInfo {
  /** 行数 */
  rowCount: number;
  /** 実行時間（ミリ秒） */
  executionTime: number | null;
  /** 最終実行日時 */
  lastExecutedAt: Date | null;
}

/**
 * クエリ実行結果
 */
export interface QueryResult {
  /** カラム情報 */
  columns: {
    name: string;
    type: string;
  }[];
  /** 行データ */
  rows: any[];
  /** クエリ情報 */
  info: QueryInfo;
}

File: src-tauri/src/models/query.rs
use serde::{Deserialize, Serialize};

/// クエリモデル（完全版）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryModel {
    /// クエリID（保存時に使用）
    pub id: Option<String>,
    /// クエリ名（保存時に使用）
    pub name: Option<String>,
    /// クエリの説明（保存時に使用）
    pub description: Option<String>,
    /// 接続ID
    pub connection_id: String,
    /// SELECT句
    pub select: SelectClause,
    /// FROM句
    pub from: FromClause,
    /// JOIN句
    pub joins: Vec<JoinClause>,
    /// WHERE句
    #[serde(default)]
    pub where_clause: Option<WhereClause>,
    /// GROUP BY句
    #[serde(default)]
    pub group_by: Option<GroupByClause>,
    /// HAVING句
    #[serde(default)]
    pub having: Option<HavingClause>,
    /// ORDER BY句
    #[serde(default)]
    pub order_by: Option<OrderByClause>,
    /// LIMIT/OFFSET
    #[serde(default)]
    pub limit: Option<LimitClause>,
    /// 作成日時
    pub created_at: Option<String>,
    /// 更新日時
    pub updated_at: Option<String>,
}

/// SELECT句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelectClause {
    /// DISTINCT
    pub distinct: bool,
    /// 選択カラム
    pub columns: Vec<SelectColumn>,
}

/// 選択カラム
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum SelectColumn {
    /// 通常のカラム
    #[serde(rename = "column")]
    Column {
        #[serde(rename = "tableAlias")]
        table_alias: String,
        #[serde(rename = "columnName")]
        column_name: String,
        alias: Option<String>,
    },
    /// 式
    #[serde(rename = "expression")]
    Expression {
        expression: String,
        alias: Option<String>,
    },
    /// 集計関数
    #[serde(rename = "aggregate")]
    Aggregate {
        aggregate: AggregateFunction,
        alias: Option<String>,
    },
    /// すべてのカラム (table.*)
    #[serde(rename = "all")]
    All {
        #[serde(rename = "tableAlias")]
        table_alias: String,
    },
}

/// 集計関数
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AggregateFunction {
    /// 関数名
    pub function: String,
    /// 対象カラム
    pub column: AggregateColumn,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AggregateColumn {
    All,
    Column {
        #[serde(rename = "tableAlias")]
        table_alias: String,
        #[serde(rename = "columnName")]
        column_name: String,
    },
}

/// FROM句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FromClause {
    /// メインテーブル
    pub table: TableReference,
}

/// テーブル参照
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableReference {
    /// スキーマ名
    pub schema: String,
    /// テーブル名
    pub name: String,
    /// エイリアス
    pub alias: String,
}

/// JOIN句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinClause {
    /// JOIN ID
    pub id: String,
    /// JOIN種別
    #[serde(rename = "type")]
    pub join_type: String,
    /// 結合テーブル
    pub table: TableReference,
    /// 結合条件
    pub conditions: Vec<JoinCondition>,
    /// 条件の結合方法
    pub condition_logic: String,
}

/// JOIN条件
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinCondition {
    /// 左側（元テーブル）
    pub left: JoinConditionColumn,
    /// 演算子
    pub operator: String,
    /// 右側（結合テーブル）
    pub right: JoinConditionColumn,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinConditionColumn {
    pub table_alias: String,
    pub column_name: String,
}

/// WHERE句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WhereClause {
    /// 条件の結合方法
    pub logic: String,
    /// 条件リスト
    pub conditions: Vec<WhereConditionItem>,
}

/// WHERE条件アイテム（条件またはグループ）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum WhereConditionItem {
    #[serde(rename = "condition")]
    Condition(WhereCondition),
    #[serde(rename = "group")]
    Group(WhereConditionGroup),
}

/// WHERE条件
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WhereCondition {
    /// 条件ID
    pub id: String,
    /// カラム
    pub column: WhereConditionColumn,
    /// 演算子
    pub operator: String,
    /// 値
    pub value: WhereValue,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WhereConditionColumn {
    pub table_alias: String,
    pub column_name: String,
}

/// WHERE値
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum WhereValue {
    #[serde(rename = "literal")]
    Literal { value: LiteralValue },
    #[serde(rename = "list")]
    List { values: Vec<LiteralValue> },
    #[serde(rename = "range")]
    Range {
        from: LiteralValue,
        to: LiteralValue,
    },
    #[serde(rename = "column")]
    Column {
        table_alias: String,
        column_name: String,
    },
    #[serde(rename = "subquery")]
    Subquery { query: Box<QueryModel> },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum LiteralValue {
    Null,
    String(String),
    Number(f64),
    Boolean(bool),
}

/// 条件グループ
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WhereConditionGroup {
    /// グループID
    pub id: String,
    /// グループ内の結合方法
    pub logic: String,
    /// 条件リスト
    pub conditions: Vec<WhereConditionItem>,
}

/// GROUP BY句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupByClause {
    /// グループ化カラム
    pub columns: Vec<GroupByColumn>,
}

/// グループ化カラム
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupByColumn {
    pub table_alias: String,
    pub column_name: String,
}

/// HAVING句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HavingClause {
    /// 条件の結合方法
    pub logic: String,
    /// 条件リスト
    pub conditions: Vec<HavingCondition>,
}

/// HAVING条件
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HavingCondition {
    /// 条件ID
    pub id: String,
    /// 集計関数
    pub aggregate: AggregateFunction,
    /// 演算子
    pub operator: String,
    /// 値
    pub value: LiteralValue,
}

/// ORDER BY句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrderByClause {
    /// ソート項目
    pub items: Vec<OrderByItem>,
}

/// ソート項目
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrderByItem {
    pub table_alias: String,
    pub column_name: String,
    pub direction: String,
    pub nulls: Option<String>,
}

/// LIMIT/OFFSET句
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LimitClause {
    /// 取得件数
    pub limit: u64,
    /// オフセット
    pub offset: Option<u64>,
}
